lib/Language/Souffle/Marshal.hs

```diff
@@ -3,50 +3,49 @@
 {-# LANGUAGE DerivingVia, TypeFamilies #-}
 {-# LANGUAGE FlexibleInstances, FlexibleContexts, DataKinds #-}
 {-# LANGUAGE UndecidableInstances, DefaultSignatures #-}
-{-# LANGUAGE ScopedTypeVariables, TypeOperators #-}
+{-# LANGUAGE ScopedTypeVariables, TypeOperators, RankNTypes #-}
 
 -- | This module exposes a uniform interface to marshal values
 --   to and from Souffle Datalog. This is done via the 'Marshal' typeclass
 --   and 'MarshalT' monad transformer.
 --   Also, a mechanism is exposed for generically deriving marshalling
 --   and unmarshalling code for simple product types.
 module Language.Souffle.Marshal
-  ( MarshalT
-  , runMarshalT
-  , Marshal(..)
+  ( Marshal(..)
+  , MarshalF(..)
+  , MarshalM
+  , interpret
   ) where
 
-import Control.Monad.Reader
-import Control.Monad.Writer
-import Control.Monad.State
-import Control.Monad.Except
-import Control.Monad.RWS
+import Control.Monad.Free
 import GHC.Generics
-import Foreign.Ptr
 import Data.Int
 import qualified Data.Text as T
 import qualified Data.Text.Lazy as TL
-import qualified Language.Souffle.Internal as Internal
 import qualified Language.Souffle.Internal.Constraints as C
 
 
-type Tuple = Ptr Internal.Tuple
 
--- | A monad transformer, used solely for marshalling and unmarshalling
---   between Haskell and Souffle Datalog.
-newtype MarshalT m a = MarshalT (ReaderT Tuple m a)
-  deriving ( Functor, Applicative, Monad
-           , MonadIO, MonadReader Tuple, MonadWriter w
-           , MonadState s, MonadRWS Tuple w s, MonadError e )
-  via ( ReaderT Tuple m )
-  deriving MonadTrans via (ReaderT Tuple)
+data MarshalF a
+  = PopInt  (Int32  -> a)
+  | PopStr  (String -> a)
+  | PushInt Int32  a
+  | PushStr String a
 
--- | Execute the monad transformer and return the result.
---   The tuple that is passed in will be used to marshal the data back and forth.
-runMarshalT :: MarshalT m a -> Tuple -> m a
-runMarshalT (MarshalT m) = runReaderT m
-{-# INLINABLE runMarshalT #-}
+instance Functor MarshalF where
+  fmap f m = case m of
+    PopInt g -> PopInt (f . g)
+    PopStr g -> PopStr (f . g)
+    PushInt i x -> PushInt i (f x)
+    PushStr s x -> PushStr s (f x)
 
+type MarshalM a = Free MarshalF a
+
+cmd :: Functor f => f a -> Free f a
+cmd command = Free (fmap Pure command)
```

Comment:
This is another way of writing `liftFree/liftF` :smile: I think we can just remove `cmd` and use `liftF` instead.